### PR TITLE
Change Dev Addr Cache to be updated with a cron function instead of a worker job

### PR DESCRIPTION
**Pull request by Mandur**
Currently the Dev Addr cache sync is working as a background job. This model is not appropriate with the Azure function and triggers errors (#258 ) in a non deterministic fashion.
Proposal is to change the hosted service to a function triggered every 5 minutes. This change would need no logic change on the current application logic.

### CODE REVIEW

File: LoRaEngine/LoraKeysManagerFacade/SyncDevAddrCache.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+        public SyncDevAddrCache(LoRaDevAddrCache loRaDevAddrCache, RegistryManager registryManager)
+        {
+            this.loRaDevAddrCache = loRaDevAddrCache;
+            this.registryManager = registryManager;
+        }
+
+        [FunctionName("SyncDevAddrCache")]
+        public async Task Run([TimerTrigger("0 */5 * * * *")]TimerInfo myTimer, ILogger log)
+        {
+            log.LogDebug($"SyncDevAddrCache function executed at: {DateTime.Now}");
```
Comment (fbeltrao) [inline LoRaEngine/LoraKeysManagerFacade/SyncDevAddrCache.cs:27]: Prefer DateTime.UtcNow as it is deterministic regardless of the host.

Comment (fbeltrao) [inline LoRaEngine/LoraKeysManagerFacade/SyncDevAddrCache.cs:27]: Wonder if we should include `myTimer` to the logs. It has [interesting information](https://docs.microsoft.com/azure/azure-functions/functions-bindings-timer?tabs=csharp#usage)

File: LoRaEngine/LoraKeysManagerFacade/SyncDevAddrCache.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        private RegistryManager registryManager;
+
+        public SyncDevAddrCache(LoRaDevAddrCache loRaDevAddrCache, RegistryManager registryManager)
+        {
+            this.loRaDevAddrCache = loRaDevAddrCache;
+            this.registryManager = registryManager;
+        }
+
+        [FunctionName("SyncDevAddrCache")]
+        public async Task Run([TimerTrigger("0 */5 * * * *")]TimerInfo myTimer, ILogger log)
```
Comment (fbeltrao) [inline LoRaEngine/LoraKeysManagerFacade/SyncDevAddrCache.cs:25]: Should we add tests to the function to ensure the device cache:
-  is being cleared 
- operation is waited
- exception are raised

?

**Review by fbeltrao — APPROVED** [review]

File: LoRaEngine/LoraKeysManagerFacade/SyncDevAddrCache.cs (hunk 1 above)
Comment (ronniesa) [reply LoRaEngine/LoraKeysManagerFacade/SyncDevAddrCache.cs:27]: Sounds like a good idea to have the timer info on the logs

**Review by ronniesa — APPROVED** [review]

File: LoRaEngine/LoraKeysManagerFacade/SyncDevAddrCache.cs (hunk 2 above)
Comment (Mandur) [reply LoRaEngine/LoraKeysManagerFacade/SyncDevAddrCache.cs:25]: I think a lot of tests regarding cache function are already present [here](https://github.com/Azure/iotedge-lorawan-starterkit/blob/fdbc3c17bf5bb0805bbf6578d8bd6d60f229395b/LoRaEngine/test/LoraKeysManagerFacade.Test/DevAddrCacheTest.cs).
Regarding the operation is waited, what exactly do you mean? And what exception do you have in mind?

File: LoRaEngine/LoraKeysManagerFacade/SyncDevAddrCache.cs (hunk 1 above)
Comment (Mandur) [reply LoRaEngine/LoraKeysManagerFacade/SyncDevAddrCache.cs]: I added some information to the log, let me know if you think it's enough

File: LoRaEngine/LoraKeysManagerFacade/SyncDevAddrCache.cs (hunk 2 above)
Comment (fbeltrao) [reply LoRaEngine/LoraKeysManagerFacade/SyncDevAddrCache.cs:25]: If loRaDevAddrCache.PerformNeededSyncs throws an exception

**Review by fbeltrao — APPROVED** [review]

**Pull request merged**